File: Selenium/Activities/Activity21.py
import win32clipboard
from selenium import webdriver
from selenium.webdriver import ActionChains, Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.firefox.service import Service as FirefoxService
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.select import Select
from selenium.webdriver.support.wait import WebDriverWait
from webdriver_manager.firefox import GeckoDriverManager

service = FirefoxService(GeckoDriverManager().install())

with webdriver.Firefox(service=service) as driver:
    wait = WebDriverWait(driver, 10)
    driver.get("https://www.training-support.net/selenium/tab-opener")

    print("Title of Multiple Tabs page: ", driver.title)

    print("Current window handle: ", driver.current_window_handle)

    driver.find_element(By.ID, "launcher").click()


    wait.until(expected_conditions.number_of_windows_to_be(2))

    print("All window handles: ", driver.window_handles)


    driver.switch_to.window(driver.window_handles[1])

    print("Current window handle: ", driver.current_window_handle)

    print("New Tab title: ", driver.title)

    heading = driver.find_element(By.XPATH, "//div[@class='content']")
    print(heading.text)

    driver.find_element(By.ID, "actionButton").click()

    wait.until(expected_conditions.number_of_windows_to_be(3))
    print("All window handles: ", driver.window_handles)


    driver.switch_to.window(driver.window_handles[2])
    print("Current window handle: ", driver.current_window_handle)

    print("New Tab title: ", driver.title)

    heading = driver.find_element(By.XPATH, "//div[@class='content']")
    print(heading.text)

    driver.close()

File: Selenium/Activities/Activity21.java
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.Alert;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.awt.datatransfer.UnsupportedFlavorException;
import java.io.IOException;
import java.time.Duration;

public class Activity21 {
    public static void main(String[] args) throws InterruptedException, IOException, UnsupportedFlavorException {
        WebDriverManager.firefoxdriver().setup();

        WebDriver driver = new FirefoxDriver();
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));

        driver.get("https://www.training-support.net/selenium/tab-opener");
        System.out.println("Title of Multiple Tabs page: " + driver.getTitle());

        System.out.println("Current tab: " + driver.getWindowHandle());
        driver.findElement(By.id("launcher")).click();

        wait.until(ExpectedConditions.numberOfWindowsToBe(2));

        System.out.println("Currently open windows: " + driver.getWindowHandles());

        for(String handle : driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }

        wait.until(ExpectedConditions.elementToBeClickable(By.id("actionButton")));
        System.out.println("Current tab: " + driver.getWindowHandle());
        System.out.println("Page title: " + driver.getTitle());
        String pageHeading = driver.findElement(By.className("content")).getText();
        System.out.println("Page Heading: " + pageHeading);
        driver.findElement(By.id("actionButton")).click();


        wait.until(ExpectedConditions.numberOfWindowsToBe(3));
        for(String handle : driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }

        wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("content")));

        System.out.println("Current tab: " + driver.getWindowHandle());

        System.out.println("Page title: " + driver.getTitle());
        pageHeading = driver.findElement(By.className("content")).getText();
        System.out.println("Page Heading: " + pageHeading);

        driver.quit();




    }
}
